File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        left = 0
        right = 0
        _hash = [-1] * 255
        result = 0

        while right < n:
            if _hash[ord(s[right])] != -1:
                if _hash[ord(s[right])] >= left:
                    left = _hash[ord(s[right])] + 1
            result = max(result, right - left + 1)
            _hash[ord(s[right])] = right 
            right += 1
        return result 

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> sett = new HashSet<>();

        int i = 0;
        int j = 0;
        int result = 0;

        while(j < s.length()) {
            Character ch = s.charAt(j);
            if(sett.contains(ch)) {
                sett.remove(s.charAt(i));
                i += 1;
            } else {
                sett.add(ch);
                result = Math.max(result, j - i + 1);
                j += 1;
            }
        }
        return result;
    }
}